### ISSUE TITLE
Running install target with root privilages breaks subsequent invocations

### ISSUE THREAD

**Issue by Psirus** [ninja-build/ninja#1302]
A "common pattern" for a library is to run `make` and then `sudo make install`. This works nicely with CMake. However, when using Ninja as the underlying build system, `sudo ninja install` changes the owner of `.ninja_log` to root. Subsequent invocations of `ninja` than fail with

     ninja: error: opening build log: Permission denied

Just deleting the file afterwards seems fine, but this is not the way to go, right? When I raised this [issue on CMake](https://gitlab.kitware.com/cmake/cmake/issues/17073), one suggestion was for Ninja to try and write the log file with the original owner if it is run as root. What do you think?

**Comment by evmar** [comment]
If you have a ninja target that builds anything, "sudo ninja ..." will build those intermediate outputs as root, which is also undesirable (the files will have the wrong owner and running the compiler as root).

I think better would be if the install commands themselves used sudo.  That is, the build file would be configured such that `ninja install` would run a command that looks like `sudo cp ...`.

**Comment by mathstuf** [comment]
Putting it into the install rules is nasty: not all installs should happen as root and not all machines use sudo.

**Comment by evmar** [comment]
Right, I meant it's up to the generator.

**Comment by Psirus** [comment]
So the CMake dev thinks this should be solved by ninja, and you think it should be solved by CMake. Understandable, and I'm not knowledgable enough to give an opinion, but unlikely to resolve the issue.

**Comment by evmar** [comment]
Can you link to the relevant cmake thread?

**Comment by mathstuf** [comment]
It's linked to in the description: https://gitlab.kitware.com/cmake/cmake/issues/17073.

**Comment by marcelhollerbach** [comment]
Hello,

How about checking if there is already a log file, if there is, ensure the permissions stay the same, if not, just rename it (and take the uid and gid of the running user)?

**Comment by ajaybd** [comment]
I faced the same issue while building the sources.
The issue is some folders are built with root permission and some other folder with other username permission.
To get rid of this, change the complete permission of the source code to a single user.
sudo chown -R username:username <android_source_code>

**Comment by jgrocha** [comment]
Same problem here. Always have to do a `chown $USER:$USER .ninja_log` after `sudo ninja install`. Thank you all for looking into this issue. You are great!

**Comment by jerome-pouiller** [comment]
It seems we can consider that as a major issue and the problem seems well understood.

So, I am surprised it is not yet fixed after 5 years. can I help on that issue? What is the problem with the proposed workaround? (a bit hacky maybe?)

**Comment by evmar** [comment]
Fundamentally when you run `sudo ninja <anything>`, ninja may write any of a bunch of files, which means those files will now be owned by root.  Not just the files that are patched in the above workaround, but also any other build intermediates.

Ninja's policy is that if you need some specific behavior for some specific task it executes, it's up to you to generate a ninja file that contains the specific task you need executed.  For that reason I think it does belong in the generator.  (Alternatively you could run your entire build as root, which seems like a bad idea in general, but that's up to you...)

**Comment by jonesmz** [comment]
> sudo ninja <anything>, ninja may write any of a bunch of files, which means those files will now be owned by root. Not just the files that are patched in the above workaround, but also any other build intermediates.

I agree with you that ninja should be expected to just do whatever is in the ninjafile with regards to file ownership of files referenced in the ninjafile.

But the files that ninja book-keeps that the ninjafile has no reference to or understanding of are quite different.

`ninja install` likely shouldn't be configured by the generator program to assume any particular user account, and per-emptively add "sudo". Instead the expected behavior that we inherit from gnu make and other similar tools is that ninja conducts operations in terms of the user it's running as, so that users who wish to install into a prefix they have write access to can do that without invoking sudo internally.

Since ninja is working with a `.ninja_*` file set that it manages, it can simply choose not to change the owner of the file when it modifies or overwrites it.

@jerome-pouiller probably what you're going to need to do, if you want to propose a code change, is modify ninja's handling of any file in the set `.ninja_*` so that if the file already exists, the new and/or updated file that `sudo ninja install` writes out be written out as owned by the original owner / group, with the original permission set if-and-only-if the current user account that ninja is running as has the ability to do that. Probably the easiest way to accomplish this is to modify the file in place, but that's most likely not a good idea because a parallel running instance of ninja might see that as a corrupted file. So you'd want to create-new, change owner/group/perms, then atomically replace the current file.

**Comment by jerome-pouiller** [comment]
For the record, the user can run `ninja -t recompact` just after `sudo ninja install`. Since the user has write access to the directory containing `.ninja_deps`, it is possible to remove `.ninja_deps` and rewrite it (exactly what  `ninja -t recompact` does).

It is probably a bit less hacky than playing with `chown`.

**Comment by jerome-pouiller** [comment]
@jonesmz your suggestion looks like the patch submitted by @marcelhollerbach. Why it has not been accepted?

**Comment by jonesmz** [comment]
Because I am not a maintainer of ninja and don't have commit / merge rights.

Not knowing which patch you're referring to (can you link the PR?) i couldn't speak to the implementation quality, but i'd be happy to review the change and offer feedback on it in the hope that it gets merged.

**Comment by jerome-pouiller** [comment]
> Not knowing which patch you're referring to (can you link the PR?)

https://github.com/marcelhollerbach/ninja/commit/1365832bf970a47c19425e4f4268fbea021c08b6

**Comment by MorelloCherry** [comment]
Hi all involved,

it has now been another year but unfortunately I can still produce this another year on. In my build I have gone to some extend using the _chown-hack_ but that is not as reliable as I have hoped.

Could I implore you to find a good solution that would work in conjunction with cmake?

It surely helps nobody if
1. The issue is being pushed to cmake and they might just do the same - we are getting nowhere with this
2. Ignoring the issue as it is - let's not pretend please the issue is not real
3. Pretending nobody cares - there are several users out there including me facing this issue

Thank you.

**Comment by am00ma** [comment]
Still facing the same issue.
Since there is an explicit `install_manifest.txt`, is it not possible to make sure sudo is used only for those files?

For reference, I think cmake/ninja can just prompt the user if the proper flags are set for the correct targets: https://stackoverflow.com/questions/46785354/sudo-during-ninja-build#comment80519603_46785890

**Comment by mathstuf** [comment]
`install_manifest.txt` is an *output* of CMake's installation support. `ninja` has nothing to do with it.
